static/js/popover_menus.js

```diff
@@ -87,4 +89,74 @@ export function initialize() {
             compose_mobile_button_popover_displayed = false;
         },
     });
+
+    function format_text(range, type) {
+        switch (type) {
+            case "bold":
+                range.start = range.start + 2;
+                range.end = range.end + 2;
+                return `**${range.text}**`;
+            case "italic":
+                range.start = range.start + 1;
+                range.end = range.end + 1;
+                return `*${range.text}*`;
+            case "link":
+                // Set cursor position inside round brackets.
+                range.end = range.end + 3;
+                range.start = range.end;
+                return `[${range.text}]()`;
+            default:
+                return range.text;
+        }
+    }
+
+    // compose formatting buttons popover
+    $("body").on("select", ".new_message_textarea, .message_edit_content ", (e) => {
+        hideAll();
+        const $textbox = $(e.currentTarget);
+        const caret_coords = util.getCaretCoordinates(e.target, e.target.selectionStart);
+        const range = $textbox.range();
+        if (!range || !range.length) {
+            // check if user selected a range of text.
+            return;
+        }
+        tippy(e.target, {
+            ...default_popover_props,
+            content: render_formatting_buttons_popover(),
+            placement: "top-start",
+            trigger: "manual",
+            // Extra 5px from top to get popover closer to text.
+            offset: [caret_coords.left, -1 * caret_coords.top + 5],
+            arrow: false,
+            showOnCreate: true,
+            onHidden(instance) {
+                instance.destroy();
+            },
+            onShow(instance) {
+                const $popper = $(instance.popper);
+                // Remove extra padding around the buttons so that
+                // user doesn't click on it by mistake.
+                $popper.find(".tippy-content").css("padding", 0);
+                $popper.on("click", ".formatting_button", (e) => {
+                    const val = $textbox.val();
+                    const pre_text = val.slice(0, range.start);
+                    const post_text = val.slice(range.end);
+                    const format_type = $(e.currentTarget).find("[data-format]").data("format");
+                    const formatted_text = format_text(range, format_type);
+                    // Replace formatted text in textbox.
+                    $textbox.val(pre_text + formatted_text + post_text);
+                    $textbox.get(0).setSelectionRange(range.start, range.end);
```

Comment:
This replacement logic should probably be a function in `compose_ui.js`; that's our common module for manipulation of the compose text areas.  (And having that logic all in one place makes it a lot easier to try to audit for bugs with breaking `Ctrl+Z` or with how whitespace is managed in the compose box, since the code for such manipulations would all be in one place).

Also, I suspect that `compose.handle_keydown` is essentially a duplicate of this code; would it make sense to extract that into functions in `compose_ui.js` and then have this and that both call the same functions?

Reply:
Oh! I wasn't aware of `compose.handle_keydown`. There is definitely code duplication here.